src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -10490,6 +10490,187 @@ public void SynthesizedDelegateTypes_NamedArguments_MoreThan16Parameters()
+                var lam5 = (S2 s2 = default) => { };
+                Report(lam5);
+                var lam6 = (int i = default) => { };
+                Report(lam6);
+
+                public struct S1 { public int X; }
+                public struct S2 { public int Y; }
+                """;
+            CompileAndVerify(source, expectedOutput: $"""
+                <>f__AnonymousDelegate0`1[System.Object]
+                <>f__AnonymousDelegate0`1[System.String]
+                <>f__AnonymousDelegate0`1[System.Nullable`1[System.Int32]]
+                <>f__AnonymousDelegate0`1[S1]
+                <>f__AnonymousDelegate0`1[S2]
+                <>f__AnonymousDelegate1`1[System.Int32]
+                """).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void SynthesizedDelegateTypes_DefaultParameterValues_IntNullableUnification()
+        {
+            var source = """
+                using System;
+                static void Report(Delegate d) => Console.WriteLine(d.GetType());
+                var lam1 = (int a = 1) => { };
+                Report(lam1);
+                var lam2 = (int? b = 1) => { };
+                Report(lam2);
+                """;
+            CompileAndVerify(source, expectedOutput: $"""
+                <>f__AnonymousDelegate0`1[System.Int32]
+                <>f__AnonymousDelegate0`1[System.Nullable`1[System.Int32]]
+                """).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void SynthesizedDelegateTypes_DefaultParameterValues_StringUnification()
+        {
+            var source = """
+                using System;
+                static void Report(Delegate d) => Console.WriteLine(d.GetType());
+                var lam1 = (string a = "") => { };
+                Report(lam1);
+                var lam2 = (string b = null) => { };
+                Report(lam2);
+                var lam3 = (string c = "abc") => { };
+                Report(lam3);
+                var lam4 = (string d = "a" + "bc") => { };
+                Report(lam4);
+                """;
+            CompileAndVerify(source, expectedOutput: $"""
+                <>f__AnonymousDelegate0`1[System.String]
+                <>f__AnonymousDelegate1`1[System.String]
+                <>f__AnonymousDelegate2`1[System.String]
+                <>f__AnonymousDelegate2`1[System.String]
+                """).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void SynthesizedDelegateTypes_ParamsArray_DifferentElementTypes()
+        {
+            var source = """
+                using System;
+                static void Report(Delegate d) => Console.WriteLine(d.GetType());
+                var lam1 = (params int[] xs) => { };
+                Report(lam1);
+                var lam2 = (params string[] ys) => { };
+                Report(lam2);
+                """;
+            CompileAndVerify(source, expectedOutput: $"""
+                <>f__AnonymousDelegate0`1[System.Int32[]]
```

Comment:
>System.Int32[] [](http://example.com/codeflow?start=42&length=14)

In metadata, ```ParamArrayAttribute``` can be applied to a parameter of a non array type. I am curious what is going to happen with a synthesized delegate type in that case. Are we going to claim the parameter is `params`? I think it is worth testing this scenario as well, perhaps in a separate test.

Reply:
Thanks, added a test.

> Are we going to claim the parameter is `params`?

Yes, the synthesized delegate was `<>f__AnonymousDelegate0<T>(params T arg)`. But after ensuring `params` synthesized delegates are templated in element type (as you suggested in another comment), the `params` modifier is dropped and `System.Action` is used instead.